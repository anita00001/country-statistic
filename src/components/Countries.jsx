import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { fetchCountryList } from '../redux/country/countrySlice';
import '../styles/Countries.css';

const Countries = () => {
  const countryList = useSelector((state) => state.country.countryList);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCountryList());
  }, [dispatch]);

  return (
    <div className="list-countries-wrapper">
      {countryList.map((item) => (
        <div key={item.name} className="list-countries">
          <h3>{item.name}</h3>
          <p>
            {item.area}
            {' '}
            kmÂ²
          </p>
        </div>
      ))}
    </div>
  );
};

export default Countries;
