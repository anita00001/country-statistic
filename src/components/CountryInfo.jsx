import { useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { GiReturnArrow } from 'react-icons/gi';
import '../styles/CountryInfo.css';

const CountryInfo = () => {
  const { name } = useParams();

  const country = useSelector((state) => state.country.countryList);
  const getCountry = country.find((c) => c.name === name);

  if (!getCountry) {
    return <div className="error">Country not found!</div>;
  }

  const recommendedCountries = () => {
    const recommend = country.filter(
      (c) => c.name !== getCountry.name
        && c.continent === getCountry.continent,
    );
    return recommend.slice(0, 5);
  };

  const getRecCountries = recommendedCountries();

  return (
    <div className="country-info-page">
      <div className="info-page-top">
        <Link to="/"><GiReturnArrow /></Link>
        <h3 className="country-name">{getCountry.name}</h3>
      </div>
      <div className="country-info-wrap">
        <img src={getCountry.flag} alt="national flag" className="national-flag" />
        <table className="info-table">
          <tbody>
            <tr>
              <td>Country</td>
              <td>{getCountry.name}</td>
            </tr>
            <tr>
              <td>Capital</td>
              <td>{getCountry.capital}</td>
            </tr>
            <tr>
              <td>Sub-region</td>
              <td>{getCountry.subregion}</td>
            </tr>
            <tr>
              <td>Region</td>
              <td>{getCountry.continent}</td>
            </tr>
            <tr>
              <td>Location</td>
              <td>
                {getCountry.location[1].toFixed(2)}
                lat,
                {' '}
                {getCountry.location[0].toFixed(2)}
                long
              </td>
            </tr>
            <tr>
              <td>Time Zone</td>
              <td>{getCountry.timezones}</td>
            </tr>
            <tr>
              <td>Population</td>
              <td>{getCountry.population}</td>
            </tr>
            <tr>
              <td>Area</td>
              <td>
                {getCountry.area}
                {' '}
                kmÂ²
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <h3 className="country-recommend">See More Countries</h3>
      <div className="recom-country-wraper">
        {getRecCountries.length === 0 ? (
          <p className="error">No recommended countries</p>
        ) : (
          getRecCountries.map((c) => (
            <div key={c.name}>
              <Link to={`/country/${c.name}`}>
                <div key={c.name} className="recom-details">
                  <p className="recom-country-name">{c.name}</p>
                  <img src={c.flag} alt="national-flag" className="recom-country-flag" />
                </div>
              </Link>
            </div>
          ))
        )}
      </div>
    </div>
  );
};

export default CountryInfo;
