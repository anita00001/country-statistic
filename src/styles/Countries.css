.country-page-top {
  position: relative;
}

.world-map-img {
  margin-left: 10%;
  width: 80%;
  height: 12rem;
}

.search-bar {
  position: absolute;
  top: 40%;
  left: 35%;
  width: 30%;
  text-align: center;
  padding: 1rem;
  outline: none;
  background: none;
  border: 1px solid #3460c7;
}

.search-bar::placeholder {
  font-size: 1rem;
  color: #2c50a6;
}

.country-page-center {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.25rem 3rem;
  border-bottom: 2px solid #3460c7;
  border-top: 2px solid #3460c7;
  background-color: #3460c74f;
}

.country-page-center > h3 {
  margin: 0;
  color: #2c50a6;
}

.country-page-center > .dropdown {
  width: 60%;
  font-size: 1rem;
  color: #2c50a6;
  padding: 0.5rem;
  outline: none;
  background-color: #f2f3f7;
  border: 1px solid #3460c7;
}

.list-countries-wrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
  margin: 1rem 1rem;
  gap: 1.5rem;
}

.list-countries {
  text-align: center;
  width: 10rem;
  height: 10rem;
  box-shadow: 0 4px 8px rgba(16, 90, 122, 0.897);
  background-color: #f2f3f7;
}

.list-countries a {
  color: #2c50a6;
  text-decoration: none;
}

.list-countries a:visited {
  color: #2c50a6;
}

.national-flag {
  margin-top: 0.2rem;
  width: 5rem;
  height: 4rem;
}

.country-name-display {
  text-align: center;
  font-size: 1rem;
  margin: -0.2rem 0;
}

.continent-name {
  font-size: 0.8rem;
  margin: -0.1rem;
}

.country-area {
  font-size: 0.8rem;
}
