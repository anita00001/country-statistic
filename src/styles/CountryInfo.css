.country-info-page {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: #fff;
  background-color: #112d71;
}

.info-page-top {
  width: 100%;
  display: flex;
  align-items: center;
  padding-left: 3rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  border-bottom: 2px solid #3460c7;
  border-top: 2px solid #3460c7;
}

.return-icon {
  color: #fff;
  font-size: 1.5rem;
}

.country-name {
  width: 50%;
  font-size: 1.2rem;
  margin: 0;
  text-align: end;
}

.country-info-wrap {
  width: 85%;
}

.country-info-wrap > .flag-wrap {
  position: relative;
  display: block;
}

.flag-wrap > .overlay {
  position: absolute;
  top: 3%;
  left: 0;
  width: 100%;
  height: 92%;
  background-color: rgba(0, 119, 255, 0.5);
}

.flag-wrap > .national-flag {
  width: 100%;
  height: 10rem;
  place-items: center;
  margin: 0.5rem 0;
  box-shadow: 0 4px 8px rgba(16, 90, 122, 0.897);
}

.info-table {
  width: 100%;
  box-shadow: 0 4px 8px rgba(16, 90, 122, 0.897);
}

.info-table > tbody > tr > td {
  padding: 0.5rem 1rem;
  width: 50%;
  font-size: 0.8rem;
  border: 2px solid #3460c7;
}

.info-table > tbody > tr:nth-child(odd) {
  background-color: #112d71;
  border: 2px solid #3460c7;
}

.info-table > tbody > tr:nth-child(even) {
  background-color: #3460c7;
  border: 2px solid #112d71;
}

.country-recommend {
  width: 80%;
  padding: 0.5rem 1.75rem;
  font-size: 1.2rem;
  margin: 1rem 1rem 0.5rem 1rem;
  background-color: #112d71;
  border: 2px solid #3460c7;
}

.recom-country-wraper {
  width: 80%;
  padding: 0 1.75rem;
  background-color: #3460c7;
  margin-bottom: 3rem;
}

.recom-country-wraper a {
  text-decoration: none;
  color: #fff;
}

.recom-details {
  margin: 0.5rem 0;
  background-color: #112d71;
}

.recom-country-name {
  display: inline-block;
  padding: 0.5rem;
  margin: 0;
}

.recom-country-flag {
  width: 1.5rem;
  height: 1rem;
}

.error {
  position: absolute;
  top: 50%;
  left: 45%;
  color: red;
  font-size: 2rem;
  text-align: center;
}
