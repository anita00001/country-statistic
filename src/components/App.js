import Countries from './Countries';
import '../styles/App.css';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <Countries />
      </header>
    </div>
  );
}

export default App;
