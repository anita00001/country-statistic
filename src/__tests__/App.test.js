import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom/extend-expect'; // for toBeInTheDocument()

import App from '../components/App';
import Countries from '../components/Countries';
import CountryInfo from '../components/CountryInfo';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  country: {
    countryList: [
      { name: 'Country 1', capital: 'Capital 1', flag: 'Flag 1', population: 1000, location: [10, 20] },
      { name: 'Country 2', capital: 'Capital 2', flag: 'Flag 2', population: 2000, location: [30, 40] },
    ],
    status: 'idle',
    error: null,
  },
};
const store = mockStore(initialState);

describe('App component', () => {
  test('renders App component: Countries component', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/']}>
          <App />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByText('Countries')).toBeInTheDocument();
  });

  test('renders App component: CountryInfo', () => {
    const countryName = 'Country 1';
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/country/${initialState.name}`]}>
          <Routes>
            <Route path="/country/:name" element={<CountryInfo getCountry={initialState} />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );

    const countryNameRegex = new RegExp(countryName, 'i');
    waitFor(() => expect(screen.getByText(countryNameRegex)).toBeInTheDocument());

    const countryCapitalRegex = new RegExp(initialState.capital, 'i');
    waitFor(() => expect(screen.getByText(countryCapitalRegex)).toBeInTheDocument());

    const countryPopulationRegex = new RegExp(initialState.population, 'i');
    waitFor(() => expect(screen.getByText(countryPopulationRegex)).toBeInTheDocument());

    const countryFlagRegex = new RegExp(initialState.flag, 'i');
    waitFor(() => expect(screen.getByText(countryFlagRegex)).toBeInTheDocument());

    const countryLocationRegex = new RegExp(initialState.location, 'i');
    waitFor(() => expect(screen.getByText(countryLocationRegex)).toBeInTheDocument());

    const countryBackButtonRegex = new RegExp('Back', 'i');
    waitFor(() => expect(screen.getByText(countryBackButtonRegex)).toBeInTheDocument());
  });

  test('renders App component: Dispatch fetchCountryList action on component mount', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </Provider>
    );

    waitFor(() => expect(store.dispatch).toHaveBeenCalledTimes(1));
  });

  test('App component matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});