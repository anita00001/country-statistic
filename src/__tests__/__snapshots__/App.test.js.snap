// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App component matches snapshot 1`] = `
<div>
  <div
    className="nav-bar-wrapper"
  >
    <p
      data-testid="time-element"
    />
    <h3>
      statistics
    </h3>
    <svg
      data-testid="microphone-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 352 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M176 352c53.02 0 96-42.98 96-96V96c0-53.02-42.98-96-96-96S80 42.98 80 96v160c0 53.02 42.98 96 96 96zm160-160h-16c-8.84 0-16 7.16-16 16v48c0 74.8-64.49 134.82-140.79 127.38C96.71 376.89 48 317.11 48 250.3V208c0-8.84-7.16-16-16-16H16c-8.84 0-16 7.16-16 16v40.16c0 89.64 63.97 169.55 152 181.69V464H96c-8.84 0-16 7.16-16 16v16c0 8.84 7.16 16 16 16h160c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16h-56v-33.77C285.71 418.47 352 344.9 352 256v-48c0-8.84-7.16-16-16-16z"
      />
    </svg>
    <svg
      data-testid="setting-icon"
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="3"
      />
      <path
        d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
      />
    </svg>
  </div>
  <section
    className="country-page"
  >
    <div
      className="country-page-top"
    >
      <img
        alt="world map"
        className="world-map-img"
        src="world-map.jpg"
      />
      <input
        className="search-bar"
        onChange={[Function]}
        placeholder="Search country by name"
        type="text"
        value=""
      />
    </div>
    <div
      className="country-page-center"
    >
      <h3>
        Countries
      </h3>
      <select
        className="dropdown"
        onChange={[Function]}
        value="All"
      >
        <option
          value="All"
        >
          All
        </option>
        <option />
      </select>
    </div>
    <div
      className="list-countries-wrapper"
    >
      <div
        className="list-countries"
        data-testid="country-item"
      >
        <a
          href="/country/Country 1"
          onClick={[Function]}
        >
          <img
            alt="national flag"
            className="national-flag"
            src="Flag 1"
          />
          <h3
            className="country-name-display"
          >
            Country 1
          </h3>
          <p
            className="continent-name"
          />
          <p
            className="country-area"
          >
             
            km²
          </p>
        </a>
      </div>
      <div
        className="list-countries"
        data-testid="country-item"
      >
        <a
          href="/country/Country 2"
          onClick={[Function]}
        >
          <img
            alt="national flag"
            className="national-flag"
            src="Flag 2"
          />
          <h3
            className="country-name-display"
          >
            Country 2
          </h3>
          <p
            className="continent-name"
          />
          <p
            className="country-area"
          >
             
            km²
          </p>
        </a>
      </div>
    </div>
  </section>
</div>
`;
